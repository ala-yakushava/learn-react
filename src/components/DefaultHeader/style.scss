@import "../../styles/theme";

.DefaultHeader {
  display: grid;
  grid-template-columns: 1fr auto;
  grid-template-rows: repeat(2, min-content);
  padding: 20px 0 150px;

  &_form {
    grid-column: 1/3;
    grid-row: 2/3;
    width: 100%;
    padding: 0 65px;
    margin: 0 auto;
    padding-top: 40px;
  }
}
