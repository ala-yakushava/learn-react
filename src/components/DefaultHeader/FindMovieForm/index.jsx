import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import { useFormik } from 'formik';

import './style.scss';
import Heading from '../../Heading';
import TextInput from '../../TextInput';
import Button from '../../Button';

const FindMovieForm = ({ className }) => {
  const formik = useFormik({
    initialValues: { text: '' },
    onSubmit: (values) => {
      console.log(values);
    },
  });

  const FormClass = cn('FindMovieForm', className);

  const {
    values, isSubmitting, dirty, isValid, handleSubmit, handleChange,
  } = formik;

  return (
    <form className={FormClass} onSubmit={handleSubmit}>
      <Heading className="FindMovieForm_title">Find your movie</Heading>
      <TextInput
        name="text"
        className="FindMovieForm_input"
        placeholder="What do you want to watch?"
        value={values.text}
        required
        onChange={handleChange}
      />
      <Button type="submit" mode="primary" disabled={isSubmitting || !dirty || !isValid}>Search</Button>
    </form>
  );
};

FindMovieForm.propTypes = {
  className: PropTypes.string,
};

FindMovieForm.defaultProps = {
  className: '',
};

export default FindMovieForm;
