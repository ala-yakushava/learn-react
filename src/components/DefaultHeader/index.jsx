import React, { useState } from 'react';

import './style.scss';
import Logo from '../Logo';
import Button from '../Button';
import FindMovieForm from './FindMovieForm';
import Modal from '../Modal';
import AddMovieForm from '../Modal/AddMovieForm';

const DefaultHeader = () => {
  const [visible, setVisible] = useState(false);

  const handleClick = () => setVisible(true);
  const handleClose = () => setVisible(false);

  return (
    <header className="DefaultHeader">
      <Logo />
      <Button mode="transparent" onClick={handleClick}>Add movie</Button>
      <FindMovieForm className="DefaultHeader_form" />
      {visible && (
        <Modal onClick={handleClose}>
          <AddMovieForm />
        </Modal>
      )}
    </header>
  );
};

export default DefaultHeader;
