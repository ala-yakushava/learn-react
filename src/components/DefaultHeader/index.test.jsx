import React from 'react';
import { useDispatch } from 'react-redux';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import DefaultHeader from './index';

jest.mock('react-redux');

describe('Components - DefaultHeader', () => {
  const dispatch = jest.fn();

  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((element) => element);
    useDispatch.mockReturnValue(dispatch);

    const modalRoot = document.createElement('div');
    modalRoot.setAttribute('id', 'modal-root');
    const body = document.querySelector('body');
    body.appendChild(modalRoot);
  });

  afterAll(() => {
    ReactDOM.createPortal.mockClear();
    useDispatch.mockClear();
  });

  test('should match snapshot', () => {
    const tree = renderer
      .create(<DefaultHeader />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should open modal with AddMovieForm', () => {
    const { getByTestId, getByRole } = render(
      <DefaultHeader />,
    );

    const leftClick = { button: 0 };
    const button = getByRole('button', { name: /Add movie/i });
    userEvent.click(button, leftClick);

    const modal = getByTestId('modal');
    const form = getByTestId('create-form');
    expect(modal).toContainElement(form);

    const closeModal = getByRole('button', { name: /X/i });
    userEvent.click(closeModal, leftClick);

    expect(modal).not.toBeInTheDocument();
  });
});
