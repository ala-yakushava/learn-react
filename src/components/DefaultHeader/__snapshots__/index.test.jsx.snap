// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - DefaultHeader should match snapshot 1`] = `
<header
  className="DefaultHeader"
  data-testid="default-header"
>
  <p
    className="Logo"
  >
    <b>
      netflix
    </b>
    roulette
  </p>
  <button
    className="Button Button--transparent"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <span
      className="Button_text"
    >
      Add movie
    </span>
  </button>
  <form
    className="FindMovieForm DefaultHeader_form"
    onSubmit={[Function]}
  >
    <div
      className="Heading FindMovieForm_title"
    >
      Find your movie
    </div>
    <div
      className="TextInput"
    >
      <input
        className="TextInput_field"
        id="text"
        name="text"
        onChange={[Function]}
        placeholder="What do you want to watch?"
        readOnly={false}
        required={true}
        type="text"
        value=""
      />
    </div>
    <button
      className="Button Button--primary"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      <span
        className="Button_text"
      >
        Search
      </span>
    </button>
  </form>
</header>
`;
